File: intellij/src/main/kotlin/dev/nylo/plugin/localizations/compare/LocaleComparator.kt
package dev.nylo.plugin.localizations.compare

import dev.nylo.plugin.localizations.model.KeyStatus
import dev.nylo.plugin.localizations.model.LocaleIssue
import dev.nylo.plugin.localizations.model.LocaleSummary
import dev.nylo.plugin.localizations.model.LocalizationReport

/**
 * Compares each locale's flattened key->value map against the [baseline]'s, producing a
 * [LocalizationReport]. Pure and IDE-free so it can be unit-tested directly.
 *
 * Mirrors Nylo's `locale:check-missing-keys`: MISSING = baseline key absent here, EMPTY = present but
 * blank, EXTRA = key here but not in baseline. [flagSameAsBase] additionally flags values identical to
 * a non-blank baseline value ([KeyStatus.UNTRANSLATED_SAME_AS_BASE]) — a heuristic, off by default.
 */
object LocaleComparator {

    fun compare(
        baseline: String,
        localeValues: Map<String, Map<String, String>>,
        parseErrors: Map<String, String> = emptyMap(),
        flagSameAsBase: Boolean = false,
    ): LocalizationReport {
        // Without a readable baseline there is nothing to compare against. Falling back to an empty
        // map would misclassify every key of every locale as EXTRA, flooding the report with bogus
        // issues. Emit summaries (the baseline row carries its parse error) but no issues at all.
        val baselineReadable = baseline in localeValues && baseline !in parseErrors
        val baseMap = if (baselineReadable) localeValues[baseline].orEmpty() else emptyMap()
        val baseKeyCount = baseMap.size

        val summaries = ArrayList<LocaleSummary>()
        val issues = ArrayList<LocaleIssue>()

        summaries += LocaleSummary(
            locale = baseline,
            isBaseline = true,
            baselineKeyCount = baseKeyCount,
            translated = baseMap.count { it.value.isNotBlank() },
            missing = 0, empty = 0, extra = 0, sameAsBase = 0,
            parseError = parseErrors[baseline],
        )

        val otherLocales = (localeValues.keys + parseErrors.keys)
            .asSequence()
            .filter { it != baseline }
            .distinct()
            .sorted()
            .toList()

        for (locale in otherLocales) {
            val parseError = parseErrors[locale]
            if (parseError != null) {
                summaries += LocaleSummary(
                    locale = locale, isBaseline = false, baselineKeyCount = baseKeyCount,
                    translated = 0, missing = 0, empty = 0, extra = 0, sameAsBase = 0, parseError = parseError,
                )
                continue
            }

            val map = localeValues[locale].orEmpty()

            if (!baselineReadable) {
                summaries += LocaleSummary(
                    locale = locale, isBaseline = false, baselineKeyCount = 0,
                    translated = map.count { it.value.isNotBlank() },
                    missing = 0, empty = 0, extra = 0, sameAsBase = 0,
                )
                continue
            }

            var translated = 0
            var missing = 0
            var empty = 0
            var sameAsBase = 0

            // Walk baseline keys in baseline (insertion) order so issues come out file-ordered.
            for ((key, baseValue) in baseMap) {
                val current = map[key]
                when {
                    current == null -> {
                        missing++
                        issues += LocaleIssue(locale, key, KeyStatus.MISSING, baseValue, null)
                    }
                    current.isBlank() -> {
                        empty++
                        issues += LocaleIssue(locale, key, KeyStatus.EMPTY, baseValue, current)
                    }
                    flagSameAsBase && baseValue.isNotBlank() && current == baseValue -> {
                        translated++ // present + non-blank still counts toward completeness
                        sameAsBase++
                        issues += LocaleIssue(locale, key, KeyStatus.UNTRANSLATED_SAME_AS_BASE, baseValue, current)
                    }
                    else -> translated++
                }
            }

            // Extra keys: present here, absent from the baseline.
            var extra = 0
            for ((key, current) in map) {
                if (!baseMap.containsKey(key)) {
                    extra++
                    issues += LocaleIssue(locale, key, KeyStatus.EXTRA, null, current)
                }
            }

            summaries += LocaleSummary(
                locale = locale, isBaseline = false, baselineKeyCount = baseKeyCount,
                translated = translated, missing = missing, empty = empty, extra = extra, sameAsBase = sameAsBase,
            )
        }

        return LocalizationReport(baseline, summaries, issues)
    }
}

File: vscode/src/core/localizations/localeComparator.ts
import { KeyStatus, LocaleIssue, LocaleSummary, LocalizationReport, makeSummary } from './model';

/**
 * Port of `dev.nylo.plugin.localizations.compare.LocaleComparator`.
 *
 * Mirrors Nylo's `locale:check-missing-keys`: MISSING = baseline key absent, EMPTY = present but
 * blank, EXTRA = key here but not in baseline. [flagSameAsBase] additionally flags values identical to
 * a non-blank baseline value (off by default — brand names/tokens are false positives).
 */
export function compareLocales(
  baseline: string,
  localeValues: Map<string, Map<string, string>>,
  parseErrors: Map<string, string> = new Map(),
  flagSameAsBase = false,
): LocalizationReport {
  // Without a readable baseline there is nothing to compare against. Falling back to an empty map
  // would misclassify every key of every locale as EXTRA, flooding the report with bogus issues.
  // Emit summaries (the baseline row carries its parse error) but no issues at all.
  const baselineReadable = localeValues.has(baseline) && !parseErrors.has(baseline);
  const baseMap = baselineReadable ? localeValues.get(baseline) ?? new Map<string, string>() : new Map<string, string>();
  const baselineKeyCount = baseMap.size;

  const summaries: LocaleSummary[] = [];
  const issues: LocaleIssue[] = [];

  summaries.push(
    makeSummary({
      locale: baseline,
      isBaseline: true,
      baselineKeyCount,
      translated: countNonBlank(baseMap),
      missing: 0,
      empty: 0,
      extra: 0,
      sameAsBase: 0,
      parseError: parseErrors.get(baseline) ?? null,
    }),
  );

  const otherLocales = [...new Set([...localeValues.keys(), ...parseErrors.keys()])]
    .filter((locale) => locale !== baseline)
    .sort(compareStrings);

  for (const locale of otherLocales) {
    const parseError = parseErrors.get(locale);
    if (parseError != null) {
      summaries.push(
        makeSummary({
          locale,
          isBaseline: false,
          baselineKeyCount,
          translated: 0,
          missing: 0,
          empty: 0,
          extra: 0,
          sameAsBase: 0,
          parseError,
        }),
      );
      continue;
    }

    const map = localeValues.get(locale) ?? new Map<string, string>();

    if (!baselineReadable) {
      summaries.push(
        makeSummary({
          locale,
          isBaseline: false,
          baselineKeyCount: 0,
          translated: countNonBlank(map),
          missing: 0,
          empty: 0,
          extra: 0,
          sameAsBase: 0,
          parseError: null,
        }),
      );
      continue;
    }

    let translated = 0;
    let missing = 0;
    let empty = 0;
    let sameAsBase = 0;

    // Walk baseline keys in baseline (insertion) order so issues come out file-ordered.
    for (const [key, baseValue] of baseMap) {
      const current = map.get(key);
      if (current === undefined) {
        missing++;
        issues.push(issue(locale, key, 'missing', baseValue, null));
      } else if (isBlank(current)) {
        empty++;
        issues.push(issue(locale, key, 'empty', baseValue, current));
      } else if (flagSameAsBase && !isBlank(baseValue) && current === baseValue) {
        translated++;
        sameAsBase++;
        issues.push(issue(locale, key, 'same_as_base', baseValue, current));
      } else {
        translated++;
      }
    }

    let extra = 0;
    for (const [key, current] of map) {
      if (!baseMap.has(key)) {
        extra++;
        issues.push(issue(locale, key, 'extra', null, current));
      }
    }

    summaries.push(
      makeSummary({
        locale,
        isBaseline: false,
        baselineKeyCount,
        translated,
        missing,
        empty,
        extra,
        sameAsBase,
        parseError: null,
      }),
    );
  }

  return { baseline, summaries, issues };
}

function issue(
  locale: string,
  key: string,
  status: KeyStatus,
  baseValue: string | null,
  currentValue: string | null,
): LocaleIssue {
  return { locale, key, status, baseValue, currentValue };
}

function countNonBlank(map: Map<string, string>): number {
  let count = 0;
  for (const value of map.values()) {
    if (!isBlank(value)) {
      count++;
    }
  }
  return count;
}

function isBlank(value: string): boolean {
  return value.trim().length === 0;
}

function compareStrings(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}
